### PR TITLE
Added tests and functions to patch a ROM

### PR DESCRIPTION
**Pull request by brownian-motion**
- adds tests that cover many different parsing edge cases
- adds (untested) methods to patch a ROM stored in a file, following the example in the docs
- adds methods to produced a patched byte buffer, given a byte array representing an arbitrary ROM

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -63,6 +70,64 @@ impl<'a> Patch<'a> {
...
+    pub fn write_to_rom(&self, rom_file: &mut File) -> std::io::Result<()> {
+        for hunk in self.hunks.iter() {
+            hunk.write_to_rom(rom_file)?
+        }
+
+        if let Some(truncation) = self.truncation() {
+            rom_file.set_len(truncation as u64)?;
+        }
+
+        return std::io::Result::Ok(());
```
Comment (euclio) [inline src/lib.rs:129]:
```suggestion
        Ok(())
```

File: src/lib.rs [hunk 2]
```diff
@@ -94,6 +163,44 @@ impl<'a> Hunk<'a> {
...
+        return res;
+    }
+
+
+    /// Writes the bytes of this hunk to the correct spot of the given binary file.
+    fn write_to_rom<T: Seek + Write>(&self, rom_file: &mut T) -> std::io::Result<()> {
+        rom_file.seek(SeekFrom::Start(self.offset() as u64))?;
+        rom_file.write_all(self.payload())?;
+
+        return std::io::Result::Ok(());
```
Comment (euclio) [inline src/lib.rs:202]:
```suggestion
        Ok(())
```

File: src/lib.rs [hunk 3]
```diff
@@ -63,6 +70,64 @@ impl<'a> Patch<'a> {
...
+
+        if let Some(truncation) = self.truncation() {
+            new_rom_data.truncate(truncation)
+        }
+
+        new_rom_data
+    }
+
+    /// Overwrites the data in the given binary file by writing each hunk and truncating if specified.
+    pub fn write_to_rom(&self, rom_file: &mut File) -> std::io::Result<()> {
```
Comment (euclio) [inline src/lib.rs:120]:
```suggestion
    pub fn write_to_rom(&self, rom_file: &mut File) -> io::Result<()> {
```

File: src/lib.rs [hunk 4]
```diff
@@ -133,3 +240,234 @@ fn be_int(len: usize) -> impl Fn(&[u8]) -> IResult<&[u8], usize> {
         Ok((input, BigEndian::read_uint(bytes, len) as usize))
     }
 }
+
+/// Simple utility to append all of the bytes in the given slice onto the given vector.
+/// This isn't super necessary but makes a lot of the code for writing files simpler.
+fn push_all(vec: &mut Vec<u8>, data: &[u8]) {
```
Comment (euclio) [inline src/lib.rs:246]: [`extend_from_slice`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.extend_from_slice) would be more idiomatic.

File: src/lib.rs [hunk 5]
```diff
@@ -63,6 +70,64 @@ impl<'a> Patch<'a> {
     pub fn truncation(&self) -> Option<usize> {
         self.truncation
     }
+
+    /// Returns the bytes that would be stored in an equivalent .ips file
+    pub fn ips_file_bytes(&self) -> Vec<u8> {
```
Comment (euclio) [inline src/lib.rs:75]: A better API might be to take an `impl Write`. Then clients would be free to pass a buffer or a file.

**Review by euclio — CHANGES_REQUESTED** [review]
Thanks for the PR! I left a few comments, and added a CI workflow.
